src/overlays/actors/ovl_En_Slime/z_en_slime.c

```diff
@@ -88,103 +128,1168 @@ static DamageTable D_80A31B1C = {
+        this->dropObjectTex = Lib_SegmentedToVirtual(gDropMagicSmallTex);
+        this->actor.hintId = TATL_HINT_ID_GREEN_CHUCHU;
+    } else if (EN_SLIME_GET_MAIN_TYPE(&this->actor) == EN_SLIME_TYPE_RED) {
+        this->dropObjectTex = Lib_SegmentedToVirtual(gDropRecoveryHeartTex);
+        this->actor.hintId = TATL_HINT_ID_RED_CHUCHU;
+    } else {
+        this->actor.hintId = TATL_HINT_ID_BLUE_CHUCHU;
+    }
+
+    // Sets actor scale and initial action func
+    EnSlime_InitBehavior(this);
+}
+
+/**
+ * Destroy EnSlime instance.
+ */
+void EnSlime_Destroy(Actor* thisx, PlayState* play) {
+    EnSlime* this = THIS;
+    Collider_DestroyCylinder(play, &this->collider);
+}
+
+/**
+ * Upon immobilization via ice arrow (for chuchu types outside of blue), initialize
+ * ice visual effects before chuchu enters immobile state.
+ */
+void EnSlime_InitIceStun(EnSlime* this) {
+    this->damageEffectType = ACTOR_DRAW_DMGEFF_FROZEN_NO_SFX;
+    this->effectScale = 0.4f;
+    this->collider.base.colType = COLTYPE_HIT3;
+    this->frozenSteamScale = 0.6f;
+    this->effectAlpha = 1.0f;
+    this->timer = 80;
+    this->actor.flags &= ~ACTOR_FLAG_400;
+    Actor_SetColorFilter(&this->actor, 0x4000, 255, 0x2000, 80);
+}
+
+/**
+ * If ice effect was recently activated (ie. non-blue chuchu was frozen by ice arrow),
+ * and the particle effect hasn't spawned yet, update the damage effect state and
+ * spawn the ice effects.
+ */
+void EnSlime_UpdateIceEffects(EnSlime* this, PlayState* play) {
+    if (this->damageEffectType == ACTOR_DRAW_DMGEFF_FROZEN_NO_SFX) {
+        this->damageEffectType = 0; // So it's not triggered again until InitIceStun has been called again.
+        this->collider.base.colType = COLTYPE_NONE;
+        this->effectAlpha = 0.0f;
+        Actor_SpawnIceEffects(play, &this->actor, this->limbPos, EN_SLIME_LIMBPOS_COUNT, 2, 0.2f, 0.2f);
+        this->actor.flags |= ACTOR_FLAG_200;
+    }
+}
+
+/**
+ * If blink animation is in progress, advance to next frame.
+ * If eyes are open, randomly (5% chance) initialize new blink cycle.
+ */
+void EnSlime_UpdateBlink(EnSlime* this) {
+    if (this->eyeTexIndex != EN_SLIME_EYETEX_OPEN) {
+        if (++this->eyeTexIndex == 4) {
+            this->eyeTexIndex = 0;
+        }
+    } else if (Rand_ZeroOne() < 0.05f) {
+        this->eyeTexIndex = EN_SLIME_EYETEX_HALF;
+    }
+}
+
+/**
+ * Set the actor's initial scale and action function.
+ */
+void EnSlime_InitBehavior(EnSlime* this) {
+    this->actor.scale.x = 0.008f;
+    this->actor.scale.z = 0.008f;
+    this->actor.scale.y = 0.011f;
+    this->actionFunc = EnSlime_InitializeIdle;
+}
+
+/**
+ * Once BGCHECK_CHECK_WALL flag is set, idle action cycle is entered.
+ * This function is only called right after the actor is first initialized.
+ * It is not set as the action function again.
+ */
+void EnSlime_InitializeIdle(EnSlime* this, PlayState* play) {
+    if (this->actor.bgCheckFlags & BGCHECK_CHECK_WALL) {
```

Comment:
```suggestion
 * Once BGCHECKFLAG_GROUND flag is set, idle action cycle is entered.
 * This function is only called right after the actor is first initialized.
 * It is not set as the action function again.
 */
void EnSlime_InitializeIdle(EnSlime* this, PlayState* play) {
    if (this->actor.bgCheckFlags & 1) {
```

`BGCHECKFLAG_GROUND` comes from the OoT repo, as MM hasn't defined flags for these yet for w/e reason.